### ISSUE TITLE
RetinaFace Confidence score and eleminate poor faces.

### ISSUE THREAD

**Issue by MyraBaba** [deepinsight/insightface#1075]
Hi,

First of all thanks for such power full face detection algorithm.

How can I a get a score for detected face confidence ? 

And is there any way to get whether landmark visible or not ? 

I attached few test images and I don't want them detected.

Any idea ? 
Best 

![Screenshot from 2020-03-09 18-14-24](https://user-images.githubusercontent.com/17505439/76230590-f49d6a80-6234-11ea-8a49-c41d9402db57.png)
![Screenshot from 2020-03-09 18-13-37](https://user-images.githubusercontent.com/17505439/76230594-f6672e00-6234-11ea-8214-0551ebb08c4d.png)
![Screenshot from 2020-03-09 18-13-28](https://user-images.githubusercontent.com/17505439/76230598-f7985b00-6234-11ea-9c33-ee1803c32885.png)

**Comment by jmorrill** [comment]
As far as I could tell, you aren't going to get a good metric from retinaface on quality, only confidence.

You can pull the 512 face features from the recognition model, then l2 norm the result to get something of a quality.  Lower numbers (~10?) will be worse quality, higher numbers will be better quality (~28?)

**Comment by MyraBaba** [comment]
do you mean the insightface mxnet features result (512) ?

and numpy.linalg.norm(feature512) . is the way ?

**Comment by jmorrill** [comment]
> do you mean the insightface mxnet features result (512) ?
> 
> and numpy.linalg.norm(feature512) . is the way ?

Yeah the feature results from arcface.

I don't know numpy or python well, but in C++ l2 norm is this:

```
    float l2_norm(const std::vector<float>& f)
    {
        float m_sum = 0.0;
        for (floati : f)
        {
            m_sum += i * i;
        }
        return std::sqrt(m_sum);
    }
```

**Comment by MyraBaba** [comment]
Thx

**Comment by pasa13142** [comment]
its out of topic but
How could you extract aligned face from image ?

**Comment by MyraBaba** [comment]
@jmorrill  Hi,

I tested 128 mobile model feature with the both eigen sqrtnorm(l2) and the above given your code

and the std::sqrt(m_sum) is always 1 or 0.99xx ?

**Comment by liutianyuan** [comment]
> @jmorrill Hi,
> 
> I tested 128 mobile model feature with the both eigen sqrtnorm(l2) and the above given your code
> 
> and the std::sqrt(m_sum) is always 1 or 0.99xx ?




> Hi,
> 
> First of all thanks for such power full face detection algorithm.
> 
> How can I a get a score for detected face confidence ?
> 
> And is there any way to get whether landmark visible or not ?
> 
> I attached few test images and I don't want them detected.
> 
> Any idea ?
> Best
> 
> ![Screenshot from 2020-03-09 18-14-24](https://user-images.githubusercontent.com/17505439/76230590-f49d6a80-6234-11ea-8a49-c41d9402db57.png)
> ![Screenshot from 2020-03-09 18-13-37](https://user-images.githubusercontent.com/17505439/76230594-f6672e00-6234-11ea-8214-0551ebb08c4d.png)
> ![Screenshot from 2020-03-09 18-13-28](https://user-images.githubusercontent.com/17505439/76230598-f7985b00-6234-11ea-9c33-ee1803c32885.png)

@MyraBaba 
I have the same issue, have you got solved？
Thank you.

**Comment by MyraBaba** [comment]
@yingfeng 
not solved yet..

**Comment by jmorrill** [comment]
> @jmorrill Hi,
> 
> I tested 128 mobile model feature with the both eigen sqrtnorm(l2) and the above given your code
> 
> and the std::sqrt(m_sum) is always 1 or 0.99xx ?

I do this on the 512 features.  I do from C++ too, so i dont know if you are doing this with the raw output of the network, or if the features have already been normalized.  (I do it with the values right from the network)  

I got the idea from page 4 of this document (top right).
https://arxiv.org/pdf/1804.01159.pdf

**Comment by MyraBaba** [comment]
What I saw in my code its normalized . So I need to do this before normalization right ? 

Best

**Comment by KanchanIIT** [comment]
does anyone solved this issue? thanks!

**Comment by avn3r** [comment]
yeah having keypoints scores (visibility) and detection scores are of great value.

**Comment by cesarandreslopez** [comment]
This would be of great value, agreed.

**Comment by MyraBaba** [comment]
@avn3r 

Would you mind to give a small example or direction for both having keypoints scores (visibility) and detection scores ?

**Comment by Arslan-Mehmood1** [comment]
> hi @MyraBaba its out of topic but How did you extract aligned face from image ?

deepface.extract_faces returns aligned faces for some detectors like retinaface
